srtcore/channel.cpp

```diff
@@ -578,20 +792,96 @@ EReadStatus CChannel::recvfrom(sockaddr* addr, CPacket& packet) const
+
+    return RST_OK;
 #endif
+}
+
+EReadStatus CChannel::recvfrom(sockaddr* addr, CPacket& packet, uint64_t uptime_us) const
+{
+    EReadStatus status = RST_OK;
+
+    // uptime_us specifies the absolute time to wait. select uses
+    // relative time. As absolute time, it might happen that the
+    // uptime is in the past.
+
+    uint64_t currtime_tk;
+    CTimer::rdtsc(currtime_tk);
+    uint64_t now = CTimer::tk2us(currtime_tk);
+    uint64_t timeout_us;
+
+    // The maximum time for poll is defined as 0.5s, so
+    // we know it can't exceed 0.99s.
+    if (uptime_us)
+    {
+        if (uptime_us <= now)
+            timeout_us = 0; // the time is already in the past, don't wait.
+        else
+        {
+            // Don't wait longer than 0.5s, even if this somehow results
+            // from calculations.
+            timeout_us = std::min(uptime_us - now, +MAX_POLL_TIME_US);
+        }
+    }
+    else
+    {
+        // Waiting forever is risky, so in case of no timeout
+        // simply wait longer time - 0.5s
+        timeout_us = MAX_POLL_TIME_US;
+    }
+
+    HLOGC(mglog.Debug, log << "CChannel::recvfrom: poll for event: usec=" << timeout_us);
+
+    m_EventRunner.poll(timeout_us);
+
+    int psg = m_EventRunner.clearSignalReading();
+    int socket_ready = m_EventRunner.socketReady();
+
+#if ENABLE_HEAVY_LOGGING
+    {
+        // The last one is rather impossible, but it is a potential value in the code.
+        static const char* type [] = {"none", "read", "error", "read+error"};
+        static const char* psgname [] = {"none", "close", "newunit", "newconn"};
+
+        const char* sock_state = type[socket_ready];
+        const char* signal_state = psgname[psg+1];
+
+        LOGC(mglog.Debug, log << "CChannel::recvfrom: poll-ready: socket="
+                << sock_state << " trigger=" << signal_state);
+    }
+
+#endif
+
+    // If the select DID NOT report read-readiness of the UDP socket,
+    // do not call the ::recvmsg function - it's useless and won't read
+    // anything, but it will still take some CPU.
+    if (!socket_ready || psg == PSG_CLOSE )
+    {
+        HLOGC(mglog.Debug, log << "CChannel::recvfrom: socket "
+                << (psg == PSG_CLOSE ? "TO BE CLOSED" : "not ready")
+                << ", possibly trigger or timeout. NOT READING.");
+        status = psg == PSG_CLOSE ? RST_ERROR : RST_AGAIN;
+        goto Return_error;
+    }
+    // NOTE: allowed to call reading in case when select reported error.
+    // This should probably repeat this same error when ::recvfrom is called.
+    // The handler of the error will decide what to do with this. It is not
+    // expected that the error be reported every time at the call.
 
+    int res, msg_flags;
+    status = sys_recvmsg(Ref(packet), Ref(res), Ref(msg_flags), addr);
 
     // Sanity check for a case when it didn't fill in even the header
-    if (size_t(recv_size) < CPacket::HDR_SIZE)
+    if (size_t(res) < CPacket::HDR_SIZE)
```

Comment:
❗️ 
 `size_t(-1)` makes this condition false.
 `status = RST_AGAIN` returned from `sys_recvmsg(…)` is not handled here.

Reply:
Forgotten status. This should jump to `Return_error` if anything else than RST_OK is returned.